collectfast/strategies/boto3.py

```diff
@@ -29,6 +49,51 @@ def _clean_hash(quoted_hash: Optional[str]) -> Optional[str]:
         assert quoted_hash[0] == quoted_hash[-1] == '"'
         return quoted_hash[1:-1]
 
+    def get_aws_hash(self, open_handle: IO) -> str:
+        """Calculate multipart-friendly hash using `multipart_chunksize` chunk size."""
+        chunk_hashes = []
+        while True:
+            data = open_handle.read(multipart_chunksize)
+            if not data:
+                break
+            chunk_hashes.append(hashlib.md5(data))
+
+        if len(chunk_hashes) < 1:
+            return hashlib.md5().hexdigest()
+
+        if len(chunk_hashes) == 1:
+            return chunk_hashes[0].hexdigest()
+
+        digests = b"".join(m.digest() for m in chunk_hashes)
+        digests_md5 = hashlib.md5(digests)
+        return "{}-{}".format(digests_md5.hexdigest(), len(chunk_hashes))
+
+    def get_gzipped_aws_hash(self, open_handle: IO) -> str:
+        """Calculate multipart-friendly gzipped hash."""
+        buffer = BytesIO()
+        zf = gzip.GzipFile(mode="wb", fileobj=buffer, mtime=0.0)
+        zf.write(force_bytes(open_handle.read()))
+        zf.close()
+        buffer.seek(0)
+        return self.get_aws_hash(buffer)
+
+    @lru_cache(maxsize=None)
+    def get_local_file_hash(self, path: str, local_storage: Storage) -> str:
+        """Calculate large file hashes differently, if necessary."""
+        # # Check if content should be gzipped and hash gzipped content
+        content_type = mimetypes.guess_type(path)[0] or "application/octet-stream"
+        with local_storage.open(path, "rb") as f:
+            file_hash = self.get_aws_hash(f)
+            if self.use_gzip and content_type in settings.gzip_content_types:
+                cache_key = self.get_cache_key("gzip_hash_%s" % file_hash)
+                gzip_file_hash = cache.get(cache_key, False)
+                if gzip_file_hash is False:
+                    f.seek(0)
+                    gzip_file_hash = self.get_gzipped_aws_hash(f)
+                    cache.set(cache_key, file_hash)
+                return str(gzip_file_hash)
+        return str(file_hash)
```

Comment:
It'd be nice if we can flatten this logic slightly. Will this work?

```suggestion
        with local_storage.open(path, "rb") as f:
            file_hash = str(self.get_aws_hash(f))
            if not self.use_gzip or content_type not in settings.gzip_content_types:
                return file_hash
            cache_key = self.get_cache_key("gzip_hash_%s" % file_hash)
            gzip_file_hash = cache.get(cache_key, False)
            if gzip_file_hash is False:
                f.seek(0)
                gzip_file_hash = self.get_gzipped_aws_hash(f)
                cache.set(cache_key, file_hash)
            return str(gzip_file_hash)
```

Reply:
Much nicer, thank you.